src/main/java/uk/gov/ch/oauth/Oauth2.java

```diff
@@ -142,29 +136,29 @@ private String getSessionNonce() {
         return oauth2Nonce;
     }
 
-    public UserProfileResponse getUserProfile(final String code, final Session chSession) {
-        LOGGER.debug("Requesting User Profile");
-
-        final OAuthToken oauthToken = getOAuthToken(code);
-        final WebClient webClient = WebClient.create();
-        final URI profileUrl = URI.create(identityProvider.getProfileUrl());
-
-        final UserProfileResponse userProfile =
-                getUserProfileResponse(oauthToken, webClient, profileUrl);
+    private UserProfileResponse fetchUserProfile(final String code) {
+        final OAuthToken oauthToken = requestOAuthToken(code);
 
+        UserProfileResponse userProfile = requestUserProfile(oauthToken);
+        if ((userProfile.getId() == null) || userProfile.getId().isEmpty()) {
```

Comment:
Not sure this is the right way to go about implementing error handling for the userprofile. I think we need to look at implementing error handing directly on the webclient. Can we undo this please as this doesn't satisfy FA-706.

Reply:
The blocking webclient call `userProfileResponse.block(timeoutDuration)` made in `requestUserProfile` will return an empty UserProfile (o throw an exception if it fails). The original code that was refactored without change, checked to see if `userProfile` as `null`. Intellij pointed out that this could never occur, so this is a way that will deal with that problem.

Reply:
To extend on this, if we make a call where there is no matching response we might expect to receive an empty JSON object. However Jackson might deserialise that as a real object with null values in all fields. This check is explicitly to handle that scenario.

Reply:
I agree this provides a useful safety net, however I still think we would need to implement checks directly on the web client for error statuses such as access denied & failure. The perl impl handles these and then redirect so I still think some work needs to be done, which of course can be picked as part of it's own ticket.

Reply:
They are trapped in the controller, as failures, including timeouts generate exceptions, but as an async web client is used, this will only occur during the `block` statement.